cpp/arcticdb/pipeline/write_frame.cpp

```diff
@@ -33,142 +33,195 @@ namespace arcticdb::pipelines {
+                    opt_error->raise(frame_->desc.fields(0).name(), offset_in_frame);
                 }
             }
 
-            for (size_t col = 0, end = slice.col_range.diff(); col < end; ++col) {
-                auto abs_col = col + frame.desc.index().field_count();
-                auto &fd = slice.non_index_field(col);
-                auto &tensor = frame.field_tensors[slice.absolute_field_col(col)];
+            for (size_t col = 0, end = slice_.col_range.diff(); col < end; ++col) {
+                auto abs_col = col + frame_->desc.index().field_count();
+                auto& fd = slice_.non_index_field(col);
+                auto& tensor = frame_->field_tensors[slice_.absolute_field_col(col)];
                 auto opt_error = aggregator_set_data(
                     fd.type(),
-                    tensor, agg, abs_col, rows_to_write, offset_in_frame, slice_num_for_column,
-                    regular_slice_size, sparsify_floats);
+                    tensor, agg, abs_col, rows_to_write, offset_in_frame, slice_num_for_column_,
+                    regular_slice_size, sparsify_floats_);
                 if (opt_error.has_value()) {
                     opt_error->raise(fd.name(), offset_in_frame);
                 }
             }
 
-            ++slice_num_for_column;
             agg.end_block_write(rows_to_write);
             agg.commit();
-        }
-    });
+            return output;
+        });
+    }
+};
+
+folly::Future<std::vector<SliceAndKey>> write_slices(
+        const std::shared_ptr<InputTensorFrame> &frame,
+        std::vector<FrameSlice>&& slices,
+        const SlicingPolicy &slicing,
+        IndexPartialKey&& key,
+        const std::shared_ptr<stream::StreamSink>& sink,
+        const std::shared_ptr<DeDupMap>& de_dup_map,
+        bool sparsify_floats) {
+    ARCTICDB_SAMPLE(WriteSlices, 0)
 
-    auto fut_writes = sink->batch_write(std::move(key_segs), de_dup_map);
+    std::vector<std::pair<FrameSlice, size_t>> slice_and_rowcount;
+    slice_and_rowcount.reserve(slices.size());
+    size_t slice_num_for_column = 0;
+    std::optional<size_t> first_row;
+    for(const auto& slice : slices) {
+        if (!first_row)
+            first_row = slice.row_range.first;
 
-    ARCTICDB_SUBSAMPLE_DEFAULT(WriteSlicesWait)
-    return std::move(fut_writes).thenValue([slices = std::move(slices)](auto &&keys) mutable {
-        std::vector<SliceAndKey> res;
-        res.reserve(keys.size());
-        for (std::size_t i = 0; i < res.capacity(); ++i) {
-            res.emplace_back(SliceAndKey{slices[i], std::move(to_atom(keys[i]))});
-        }
-        return res;
-    });
+        if (slice.row_range.first == first_row.value())
+            slice_num_for_column = 0;
+
+        slice_and_rowcount.emplace_back(slice, slice_num_for_column);
+        ++slice_num_for_column;
+    }
+
+    const auto write_window = ConfigsMap::instance()->get_int("VersionStore.BatchWriteWindow", 50);
```

Comment:
Can we default to something less heurisitic, like `2 * io_thread_count()`? We also need to be careful of the max task queue size for the CPU executor in `batch_write` situations though.